File: 2025/133. Clone Graph/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {

    public Node cloneGraph(Node node) {
        Map<Integer, Node> map = new HashMap<>();
        return traverse(node, map);
    }

    private Node traverse(Node node, Map<Integer, Node> map) {
        if (node == null)
            return node;
        if (map.containsKey(node.val)) 
            return map.get(node.val);
        Node newnode = new Node(node.val);
        map.put(node.val, newnode);
        newnode.neighbors = new ArrayList<>();
        for (Node n : node.neighbors) {
            Node tmp = traverse(n, map);
            newnode.neighbors.add(tmp);
        }
        return newnode;
    }
}

File: 2025/133. Clone Graph/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        graph_map = {}

        def dfs_traverse(node: Optional['Node']) -> Optional['Node']:
            if node:
                clone_node = Node(node.val)
                graph_map[node] = clone_node
                for n in node.neighbors:
                    clone_node.neighbors.append(graph_map[n]) if n in graph_map else clone_node.neighbors.append(dfs_traverse(n))
                return clone_node

        return dfs_traverse(node)